earthly/rust/Earthfile

```diff
@@ -0,0 +1,75 @@
+# Common Rust UDCs and Builders.
+VERSION 0.7
+
+# cspell: words rustup automake autotools xutils miri nextest kani
+
+# Base Rustup build container.
+rustup:
+    # Base it on Debian bookworm-slim
+    FROM debian:bookworm-slim
+
+    WORKDIR /root
+
+    # Install necessary packages 
+    # Expand this list as needed, rather than adding more tools in later containers.
+    RUN apt-get update && \
+        apt-get install --no-install-recommends -y \
+            ca-certificates \
+            curl \
+            file \
+            build-essential \
+            autoconf \
+            automake \
+            autotools-dev \
+            libtool \
+            xutils-dev \
+            mold \
+            python3 \
+            python3-pip \
+            musl-tools && \
+        rm -rf /var/lib/apt/lists/*
+
+    # Install OpenSSL here but ONLY if we can't avoid using it.
+
+    # Install rustup.
+    RUN curl https://sh.rustup.rs -sSf | sh -s -- --default-toolchain none -y -v
+
+    # Set necessary env vars from this setup.
+    ENV PATH=/root/.cargo/bin:$PATH
+
+# Common Rust setup
+# Parameters:
+#  * toolchain : The `rust-toolchain` toml file.
+RUST_SETUP:
+    COMMAND
+    ARG toolchain
+
+    # Poetry Installation directory.
+    # Gives poetry and our poetry project a reliable location.
+    WORKDIR /build
+
+    # Copy our toolchain dependency.
+    COPY $toolchain ./rust-toolchain.toml
```

Comment:
I think you _might_ be duplicating work. IIRC, the `rustup` script will install a default version of Rust as the initial toolchain. I think there's a way to have it instead installed from a toolchain file.  It's a slight optimization, as it avoids installing Rust twice.

Reply:
There isn't.  I went through that.  I specifically don't install a default toolchain when I install rustup, just rustup itself.
Ideally we wouldn't be changing the base os + rustup build very often once we have a mature build process.


So the idea was, just install `rustup` in the base os image and no rust version at all.
Then install the rust version defined by the `rust-toolchain.toml` itself.
Thats why I used `--default-toolchain none` on the rustup install.

The way rustup works is it installs the latest rust by default, or a specific named version, ie `1.72` or `stable` or `nightly`, etc.
`stable` always installs the latest version, which we may not need or use, especially if the base-os is a couple of rust releases behind (which is possible), so the latest rust when the `rustup` layer was built in CI, may not be the latest version by the time we use it to do a build.  So installing one then seems redundant.  It also doesn't mark the version it installs from `rust-toolchain.toml` as a default and there was no way I could find to get it to do so.  

I think the intention is `rustup` installs many SYSTEM versions of rust.  One can be manually set to the SYSTEM default, but if a `rust-toolchain.toml` is present, it will use that version and install it if asked, but it doesn't;t make it default.  I had to jump through hoops to get it to make the `rust-toolchain.toml` actually be the default by extracting the version from it and manually telling `rustup` to do it.  its not really designed to be ergonomic for a container build use case at all.

`rustup` has no facility to tell it to "use" a particular `rust-toolchain.toml` it just picks it up if its present in the CWD.  And if you said `rustup install stable` it ignores it completely and installs the latest stable rust.  Which isn't what we want.

All said `rustup` is a PITA, but I have worked around its shortcomings.

Reply:
Yeah, I had a similar experience, but it gets the job done I guess.